patch/driver-support-optoe-write-max.patch

```diff
@@ -0,0 +1,46 @@
+From 83573e4c3171b528a4516caf3483cfccf90b9a32 Mon Sep 17 00:00:00 2001
+From: Prince George <prgeor@microsoft.com>
+Date: Wed, 4 Aug 2021 10:06:18 +0000
+Subject: [PATCH] Increase optoe's write-max to support i2c block write
+
+In current optoe driver, the value of write_max is hardcoded to one byte.
+CMIS spec supports firmware upgrade on QSFP-DD transceivers which means
+there is significant overhead due to this one byte write limit when a
+platform wants to upgrade the firmware over an i2c bus at 100Khz. The
+overhead is more pronounced when a platform has shared master and
+platform needs to perform firmware upgrade on multiple transceivers.
+
+Signed-off-by: Prince George <prgeor@microsoft.com>
+---
+ drivers/misc/eeprom/optoe.c | 7 ++++---
+ 1 file changed, 4 insertions(+), 3 deletions(-)
+
+diff --git a/drivers/misc/eeprom/optoe.c b/drivers/misc/eeprom/optoe.c
+index 16287fdc5..6cf61e645 100644
+--- a/drivers/misc/eeprom/optoe.c
++++ b/drivers/misc/eeprom/optoe.c
+@@ -1069,7 +1069,7 @@ static int optoe_probe(struct i2c_client *client,
+ 		 * 2 byte writes are acceptable for PE and Vout changes per
+ 		 * Application Note AN-2071.
+ 		 */
+-		unsigned int write_max = 1;
++		unsigned int write_max = 32;
```

Comment:
So the decision here is to hardcode the `write_max` size to SMBus 2.x max block size?
I am mostly fine with this solution but the lack of configuration knob might lead to some xcvrs not being supported by optoe and might therefore make some users unhappy.

I think one of 2 mechanism could be used to allow users with known bad xcvrs to still have them working properly.

1) Adding a module parameter that can be tweaked at runtime and `write_max` would use this value.
This would offer a global knob for the module, but changing the `write_max` speed would impact all xcvrs even though only one might need this workaround.

2) Adding a per xcvr sysfs entry called `max_write_size` similarily to the `eeprom` and `port_name` entries that already exists.

In both cases above the choice of the default value should be chosen
 - default value of 32 because there is no reason to penalize the performance of xcvr models that follow the spec
 - default value of 1 for backward compatibility and likely enhanced reliability
In both these cases it still means that some logic needs to run somewhere (xcvrd) to change the `write_max` value based on what is seen.
AFAIK write speed doesn't matter that much on xcvrs for anything else than firmware upgrades, so potentially the logic performing the upgrade could upgrade write_size to 32.

But IMO the burden of adding quirks should be on crappy devices that don't implement the spec correctly.
I don't see why everyone else should suffer from the shortcomings of a brand/model.

Reply:
Thanks Samuel for reviewing. I will wait for the performance numbers  before making the write-max configurable.

Reply:
Fixed in latest commit